File: apps/mobile/lib/features/routing/domain/route.dart
import 'package:equatable/equatable.dart';

import '../../places/domain/place.dart';

/// A calculated path between two points.
///
/// A `Route` is not a `Trail`: it is what the road network says about getting from
/// A to B, with no name, no author and no stops. A Trail is a curated experience
/// built *over* routes and places, and it belongs to a later PR (constitution
/// §Routing).
class TrilhaRoute extends Equatable {
  const TrilhaRoute({
    required this.origin,
    required this.destination,
    required this.geometry,
    required this.distanceMeters,
    required this.durationSeconds,
    required this.bounds,
    required this.legs,
  });

  final RouteEndpoint origin;
  final RouteEndpoint destination;

  /// The line to draw, in WGS84 degrees.
  final List<LatLng> geometry;

  /// Canonical units come from the backend; formatting happens in the widgets (§15).
  final int distanceMeters;
  final int durationSeconds;

  final MapBounds bounds;
  final List<RouteLeg> legs;

  @override
  List<Object?> get props => <Object?>[
    origin,
    destination,
    geometry,
    distanceMeters,
    durationSeconds,
    bounds,
    legs,
  ];
}

/// One end of a route.
///
/// [placeId] is present when the point came from a Place, and absent when it came
/// from the device's location or a tap on the map. Routing never requires one (§7).
class RouteEndpoint extends Equatable {
  const RouteEndpoint({required this.position, this.placeId, this.label});

  final LatLng position;
  final String? placeId;

  /// What to show the user — a Place name, "Your location", or coordinates.
  final String? label;

  RouteEndpoint copyWith({LatLng? position, String? placeId, String? label}) =>
      RouteEndpoint(
        position: position ?? this.position,
        placeId: placeId ?? this.placeId,
        label: label ?? this.label,
      );

  @override
  List<Object?> get props => <Object?>[position, placeId, label];
}

class RouteLeg extends Equatable {
  const RouteLeg({
    required this.distanceMeters,
    required this.durationSeconds,
    this.summary,
  });

  final int distanceMeters;
  final int durationSeconds;
  final String? summary;

  @override
  List<Object?> get props => <Object?>[
    distanceMeters,
    durationSeconds,
    summary,
  ];
}

/// Formats a duration for display (§15, §39).
///
/// The backend deals only in seconds; deciding that 6,480 reads as "1h48" is a
/// presentation choice and lives here.
String formatDuration(int seconds) {
  if (seconds < 60) return '${seconds}s';

  final int totalMinutes = (seconds / 60).round();
  final int hours = totalMinutes ~/ 60;
  final int minutes = totalMinutes % 60;

  if (hours == 0) return '${minutes}min';
  if (minutes == 0) return '${hours}h';
  return '${hours}h${minutes.toString().padLeft(2, '0')}';
}

/// Formats a route distance for display (§15).
///
/// Distinct from the Places formatter: a route is typically long enough that metres
/// are noise, so this switches to kilometres sooner and drops the decimal earlier.
String formatRouteDistance(int metres) {
  if (metres < 1000) return '$metres m';

  final double km = metres / 1000;
  return km < 100 ? '${km.toStringAsFixed(1)} km' : '${km.round()} km';
}

File: apps/api/src/modules/routing/domain/route.ts
import { InvalidRouteGeometryError } from './routing-errors.js';

/**
 * Routing domain types.
 *
 * Plain data — no Nest, no HTTP, no provider vocabulary. A `Route` is a computed path
 * between two points and nothing more: it carries no name, no author, no stops and no
 * social meaning. Those belong to `Trail`, which is a different concept entirely
 * (constitution §Routing) and arrives in its own PR.
 */

/** A point in WGS84 degrees. */
export interface RoutePoint {
  readonly latitude: number;
  readonly longitude: number;
  /** Set when the endpoint came from an existing Place. Routing never requires one. */
  readonly placeId?: string | undefined;
}

/**
 * Route geometry as GeoJSON.
 *
 * GeoJSON rather than an encoded polyline: PostGIS reads it directly with
 * `ST_GeomFromGeoJSON`, so the corridor can be derived without a decode step, and it
 * stays readable in a payload someone is debugging. An encoded polyline is a
 * transport optimisation, not a representation of truth (ADR-0013).
 */
export interface RouteGeometry {
  readonly type: 'LineString';
  /** `[longitude, latitude]` pairs — GeoJSON axis order, not lat/lng. */
  readonly coordinates: readonly (readonly [number, number])[];
}

/**
 * Canonical units: metres and seconds (constitution §Routing).
 *
 * Kilometres and minutes are presentation choices and are made in the client.
 */
export interface RouteMetrics {
  readonly distanceMeters: number;
  readonly durationSeconds: number;
}

/** Geographic extent of the geometry, for fitting a camera. */
export interface RouteBounds {
  readonly north: number;
  readonly south: number;
  readonly east: number;
  readonly west: number;
}

/**
 * One segment of a route.
 *
 * PR-03 always produces exactly one leg, because the API accepts only origin and
 * destination. The array exists so that adding waypoints later is a change to the
 * request, not a restructuring of the result (§16).
 */
export interface RouteLeg {
  readonly distanceMeters: number;
  readonly durationSeconds: number;
  /** Human-readable endpoints as the provider named them, when it supplies them. */
  readonly summary: string | null;
}

/**
 * The area within a given distance of the route.
 *
 * Derived from the geometry by PostGIS, never returned by the provider. PR-04 uses it
 * to ask "which Places lie along this route"; PR-03 only has to produce it correctly.
 */
export interface RouteCorridor {
  readonly widthMeters: number;
  /** GeoJSON Polygon or MultiPolygon, as produced by the buffer. */
  readonly geometry: unknown;
}

/** The normalised result of a route calculation. */
export interface Route {
  readonly origin: RoutePoint;
  readonly destination: RoutePoint;
  readonly geometry: RouteGeometry;
  readonly metrics: RouteMetrics;
  readonly bounds: RouteBounds;
  readonly legs: readonly RouteLeg[];
  readonly corridor: RouteCorridor | null;
  /**
   * Which provider computed this, by name (PR-05, §18).
   *
   * Non-sensitive attribution — never a URL and never a credential. A Trail persists
   * it so that a snapshot saved today is attributable after a change of supplier.
   */
  readonly provider: string;
}

/**
 * Computes bounds from the geometry rather than trusting the provider (§17).
 *
 * Deterministic and cheap, and it cannot disagree with the line that will actually be
 * drawn — which a provider-supplied bounding box can, if it describes an alternative
 * or a simplified overview.
 */
export function boundsOf(geometry: RouteGeometry): RouteBounds {
  const [first] = geometry.coordinates;
  if (first === undefined) {
    throw new Error('Cannot compute bounds of an empty geometry');
  }

  let north = first[1];
  let south = first[1];
  let east = first[0];
  let west = first[0];

  for (const [longitude, latitude] of geometry.coordinates) {
    if (latitude > north) north = latitude;
    if (latitude < south) south = latitude;
    if (longitude > east) east = longitude;
    if (longitude < west) west = longitude;
  }

  return { north, south, east, west };
}

/**
 * Great-circle distance in metres.
 *
 * Used only to reject a request *before* it reaches a metered provider — a degenerate
 * pair, or endpoints so far apart that the input is almost certainly wrong. It is not
 * a routing distance and never appears in a result: the road distance comes from the
 * provider, and PostGIS owns every spatial measurement that matters
 * (constitution §Geography).
 */
export function greatCircleMeters(a: RoutePoint, b: RoutePoint): number {
  const earthRadiusMeters = 6_371_008.8;
  const toRadians = (degrees: number): number => (degrees * Math.PI) / 180;

  const deltaLat = toRadians(b.latitude - a.latitude);
  const deltaLng = toRadians(b.longitude - a.longitude);
  const latA = toRadians(a.latitude);
  const latB = toRadians(b.latitude);

  const h =
    Math.sin(deltaLat / 2) ** 2 + Math.cos(latA) * Math.cos(latB) * Math.sin(deltaLng / 2) ** 2;

  return 2 * earthRadiusMeters * Math.asin(Math.min(1, Math.sqrt(h)));
}

/** A LineString needs at least two positions to describe a line. */
export const MINIMUM_LINE_POSITIONS = 2;

/**
 * Validates a GeoJSON LineString position by position.
 *
 * Lives in the domain because "what counts as a usable route geometry" is a Trilha
 * invariant, not a property of whichever provider produced it — and because more than
 * one entry point needs it. PostGIS is emphatically *not* a backstop here: it accepts
 * a single-position LineString and buffers it into a circle, so a degenerate geometry
 * that slipped through would produce a corridor around a point and no error anywhere.
 */
export function parseLineString(value: unknown): RouteGeometry {
  if (typeof value !== 'object' || value === null) {
    throw new InvalidRouteGeometryError('Geometry is missing');
  }

  const candidate = value as { type?: unknown; coordinates?: unknown };
  if (candidate.type !== 'LineString') {
    throw new InvalidRouteGeometryError('Geometry is not a LineString');
  }
  if (!Array.isArray(candidate.coordinates)) {
    throw new InvalidRouteGeometryError('Geometry coordinates are not an array');
  }
  if (candidate.coordinates.length < MINIMUM_LINE_POSITIONS) {
    throw new InvalidRouteGeometryError('A LineString needs at least two positions');
  }

  const coordinates: [number, number][] = candidate.coordinates.map((position: unknown) => {
    if (!Array.isArray(position) || position.length < 2) {
      throw new InvalidRouteGeometryError('A position must be a [longitude, latitude] pair');
    }

    const [longitude, latitude] = position as unknown[];
    if (
      typeof longitude !== 'number' ||
      typeof latitude !== 'number' ||
      !Number.isFinite(longitude) ||
      !Number.isFinite(latitude) ||
      longitude < -180 ||
      longitude > 180 ||
      latitude < -90 ||
      latitude > 90
    ) {
      throw new InvalidRouteGeometryError('A position is outside valid coordinate bounds');
    }

    return [longitude, latitude];
  });

  return { type: 'LineString', coordinates };
}
